kudu/src/test/java/org/apache/apex/malhar/kudu/test/KuduClusterAvailabilityTestRule.java

```diff
@@ -0,0 +1,161 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.kudu.test;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.apex.malhar.kudu.BaseKuduOutputOperator;
+import org.apache.apex.malhar.kudu.IncrementalStepScanInputOperator;
+import org.apache.apex.malhar.kudu.KuduClientTestCommons;
+import org.apache.apex.malhar.kudu.KuduInputOperatorCommons;
+
+/**
+ * A Junit rule that helps in bypassing tests that cannot be done if the kudu cluster is not present.
+ */
+public class KuduClusterAvailabilityTestRule implements TestRule
+{
+
+  private static final transient Logger LOG = LoggerFactory.getLogger(KuduClusterAvailabilityTestRule.class);
+
+  public static final String KUDU_MASTER_HOSTS_STRING_ENV = "kudumasterhosts";
+  @Override
+  public Statement apply(Statement base, Description description)
+  {
+    KuduClusterTestContext testContext = description.getAnnotation(KuduClusterTestContext.class);
+    String masterHostsValue = System.getProperty(KUDU_MASTER_HOSTS_STRING_ENV);
+    boolean runThisTest = true; // default is to run the test if no annotation is specified
+    if ( testContext != null) {
+      if ((masterHostsValue != null) && (testContext.kuduClusterBasedTest())) {
+        runThisTest = true;
+      }
+      if ((masterHostsValue != null) && (!testContext.kuduClusterBasedTest())) {
+        runThisTest = false;
+      }
+      if ((masterHostsValue == null) && (testContext.kuduClusterBasedTest())) {
+        runThisTest = false;
+      }
+      if ((masterHostsValue == null) && (!testContext.kuduClusterBasedTest())) {
+        runThisTest = true;
+      }
+    }
+    if (runThisTest) {
+      // call the before class equivalent for real kudu cluster
+      if ((testContext != null ) && (testContext.kuduClusterBasedTest())) {
+        try {
+          initRealKuduClusterSetup(masterHostsValue);
+        } catch (Exception e) {
+          LOG.error(e.getMessage(), e);
```

Comment:
why LOG.error? If it is a real problem, propagate exception.

Reply:
Fixed.